### PR TITLE
Sl 2873 spc bulk import

### PR DESCRIPTION
**Pull request by farzana-vend**
**Context:**
With the introduction of structured product codes (SPC) a lot of the beta retailers have asked for a way to get codes into Vend in bulk. 
Jira: https://vendhq.atlassian.net/browse/SL-2873

**New:**
This introduces a new command `import-product-codes` which creates product codes and assigns them to products. 
A output csv file detailing product codes that were unsuccessful
Basic unit tests covering:
- Ensuring the csv file is read and interpreted correctly, because the headers are dynamic
- Ensure duplicate validation is checked before attempting API calls

**Changes:**
The project has been converted to use go-modules
A few new helper functions have been added to facilitate read from and writing to csv
A new custom implementation for making API requests, this was necessary for two reasons:
- Ensure we could access the status code response to determine if the request succeeded or encountered validation errors
- Ensure failures would not terminate the command entirely

### CODE REVIEW

File: commands/importProductCodes.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+	err = validateHeader(header)
+	if err != nil {
+		fmt.Println("Header validation failed")
+		return nil, err
+	}
+
+	// Ensure there are no duplicate product codes
+	err = validateProductCodeUniqueness(records)
+	if err != nil {
+		fmt.Println("Uniqueness validation failed: ", err.Error())
```
Comment (jackharrisonsherlock) [inline commands/importProductCodes.go:99]: Can we update the error message to be clear that all product codes must be unique across the entire product catalogue?

Uniqueness validation failed: All product codes must be unique across the product catalogue.

File: commands/importProductCodes.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	fmt.Println("Begin processing product codes.")
+
+	for i := 0; i < len(productCodes); i += BatchSize {
+		batchNum := i + 1
+		j := i + BatchSize
+		if j > len(productCodes) {
+			j = len(productCodes)
+		}
+		// Make the request to Vend
+		fmt.Printf("Posting: %v \n", productCodes[i:j])
```
Comment (jackharrisonsherlock) [inline commands/importProductCodes.go:172]: Do we want to return a line for every request for each product code? 

We could just remove this line and leave the "Begin processing product codes..." and then wait for the successful/failed batches output?

File: commands/importProductCodes.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+	}
+
+	// If any codes failed, export them
+	if len(failedProductCodes) > 0 {
+		err := writeOutput(failedProductCodes)
+		if err != nil {
+			fmt.Printf("\nUnsuccesssful! Failed to write ouput for %d Product Codes", len(failedProductCodes))
+			return err
+		}
+		fmt.Printf("\nFinished! Partially successful, %d batches failed", len(failedProductCodes))
```
Comment (jackharrisonsherlock) [inline commands/importProductCodes.go:205]: We should let them know that the failed batches have been outputted here so they know where to look. 

Finished! Partially successful, %d batches failed. Please checked the outputted CSV file for the failed batches.

**Review by jackharrisonsherlock — APPROVED** [review]
LGTM, Just some amendments for the returned error messages so the team know how to troubleshoot any issues.

File: commands/importProductCodes.go (hunk 2 above)
Comment (farzana-vend) [reply commands/importProductCodes.go]: Ah this was for testing, missed it in clean up.

**Pull request merged**